src/ssl.c

```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
     return ret;
 }
 
+/**
+ * Creates and returns a PKCS7 signedData structure.
+ *
+ * Inner content type is set to DATA to match OpenSSL behavior.
+ *
+ * signer - certificate to sign bundle with
+ * pkey     - private key matching signcert
+ * certs    - optional additional set of certificates to include
+ * in       - input data to be signed
+ * flags    - optional set of flags to control sign behavior
+ *
+ *    PKCS7_BINARY - Do not translate input data to MIME canonical
+ *                   format (\r\n line endings), thus preventing corruption of
+ *                   binary content.
+ *    PKCS7_TEXT - Prepend MIME headers for text/plain to content.
+ *    PKCS7_DETACHED - Set signature detached, omit content from output bundle.
+ *    PKCS7_STREAM - initialize PKCS7 struct for signing, do not read data.
+ *
+ * Flags not currently supported:
+ *    PKCS7_NOCERTS - Do not include the signer cert in the output bundle.
+ *    PKCS7_PARTIAL - Allow for PKCS7_sign() to be only partially set up,
+ *                    then signers etc to be added separately before
+ *                    calling PKCS7_final().
+ *
+ * Returns valid PKCS7 structure pointer, or NULL if an error occurred.
+ */
+PKCS7* wolfSSL_PKCS7_sign(WOLFSSL_X509* signer, WOLFSSL_EVP_PKEY* pkey,
+        WOLFSSL_STACK* certs, WOLFSSL_BIO* in, int flags)
+{
+    int err = 0;
+    WOLFSSL_PKCS7* p7 = NULL;
+    WOLFSSL_STACK* cert = certs;
+
+    WOLFSSL_ENTER("wolfSSL_PKCS7_sign");
+
+    if (flags & PKCS7_NOCERTS) {
+        WOLFSSL_MSG("PKCS7_NOCERTS flag not yet supported");
+        err = 1;
+    }
+
+    if (flags & PKCS7_PARTIAL) {
+        WOLFSSL_MSG("PKCS7_PARTIAL flag not yet supported");
+        err = 1;
+    }
+
+    if ((err == 0) && (signer == NULL || signer->derCert == NULL ||
+                       signer->derCert->length == 0)) {
+        WOLFSSL_MSG("Bad function arg, signer is NULL or incomplete");
+        err = 1;
+    }
+
+    if ((err == 0) && (pkey == NULL || pkey->pkey.ptr == NULL ||
+                       pkey->pkey_sz <= 0)) {
+        WOLFSSL_MSG("Bad function arg, pkey is NULL or incomplete");
+        err = 1;
+    }
+
+    if ((err == 0) && (in == NULL) && !(flags & PKCS7_STREAM)) {
+        WOLFSSL_MSG("input data required unless PKCS7_STREAM used");
+        err = 1;
+    }
+
+    if ((err == 0) && ((p7 = (WOLFSSL_PKCS7*)wolfSSL_PKCS7_new()) == NULL)) {
```

Comment:
Interesting -- how does this cast work? It seems like `WOLFSSL_PKCS7` and `PKCS7` are different types. I see this has been done elsewhere, too, but I'm just curious.

Reply:
I was following our existing pattern here.  I didn't check with Juliusz, but my best guess was that OpenSSL-consuming apps expect the returned struct to be `PKCS7`.  Since we already have a `PKCS7` struct of our own, he introduced a `WOLFSSL_PKCS7` which holds an inner `PKCS7`.  I think this should work as long as we **always** cast it back to the right `WOLFSSL_PKCS7` type before accessing/referencing members inside, but it does seem fragile.

Reply:
Got it. I will add @julek-wolfssl to review this just for thoroughness.

Reply:
I didn't actually write these data structures. That's how it was when I was working on this section as well. I also was concerned about this solution but since `PKCS7 pkcs7` is the first field of the `WOLFSSL_PKCS7` struct, their addresses are guaranteed to match. This means that accessing the `PKCS7 pkcs7` field as either `(WOLFSSL_PKCS7*)p7->pkcs7.content` or as `(PKCS7*)p7->content` is safe. Obviously if the other fields were to be accessed when only `sizeof(PKCS7)` memory is allocated then it would result in a segfault. Overall I left this as is as I didn't see a real need to refactor this.